### PR TITLE
✨Add Identity to Correlation Service

### LINKED ISSUE #225: 🐛 'Get Correlation' queries always include correlations from other users
#### Describe your issue

All queries for retrieving a users correlations include correlations that were started with different identities.

#### How can others reproduce the issue?

- Start a few ProcessInstances, using varying identities
- Make a request like `GetAllCorrelations` or `GetActiveCorrelations` through any of the public APIs
- See that the results will include all correlations, not just those of the requesting user

#### Possible solution

Implement a filter that makes sure the returned results will only contain those correlations that the user is authorized to see.

This filter should be placed in the `CorrelationService` of the `ProcessEngineCore` package.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by alexanderkasten**
**Changes:**

1. Refactor Correlation Queries to Filter Their Results by a Given Identity, to Ensure That Users Can Only See Their Own Correlations
2. Adjust And Add Tests to Check Whether a User Can Only Query Their Own Correlations

**Issues:**

Closes #225 

PR: #227

---

See the [examples](#example) for inspiration.

## How can others test the changes?

> Describe how others can test your changes (you can remove this section for typo fixes etc.)

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Fixing Security Issues   | `:lock:`             | 🔒     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️      |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️      |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

**Comment by alexanderkasten** [comment]
> Please add some tests for the Management API to verify that you really cannot get another users correlation.
> 
> These can be located in the same files as the existing Management API correlation tests.

@ElRaptorus 
There are already tests for this cases in 

- get_all_correlations.js
- get_by_process_instance_id.js
- get_by_process_model_id.js

at the end of the file.

File: test/2_management_api/correlations/get_active_correlations.js [hunk 1]
```diff
@@ -36,14 +36,13 @@ describe('Management API:   GET  ->  /correlations/active', () => {
...
 
-  it('should return all active correlations through the management api', async () => {
+  it('should return all active correlations for an user through the management api', async () => {
 
     const correlations = await testFixtureProvider
       .managementApiClientService
       .getActiveCorrelations(defaultIdentity);
 
     should(correlations).be.instanceOf(Array);
-    should(correlations.length).be.greaterThan(0);
```
Comment (ElRaptorus) [inline test/2_management_api/correlations/get_active_correlations.js]: Why was this removed?

File: test/2_management_api/correlations/get_all_correlations.js [hunk 2]
```diff
@@ -46,14 +46,13 @@ describe('Management API:   GET  ->  /correlations/all', () => {
...
 
-  it('should return all correlations through the management api', async () => {
+  it('should return all correlations for an user through the management api', async () => {
 
     const correlations = await testFixtureProvider
       .managementApiClientService
       .getAllCorrelations(testFixtureProvider.identities.defaultUser);
 
     should(correlations).be.instanceOf(Array);
-    should(correlations.length).be.greaterThan(0);
```
Comment (ElRaptorus) [inline test/2_management_api/correlations/get_all_correlations.js]: Why was this removed?

File: test/2_management_api/correlations/get_by_process_model_id.js [hunk 3]
```diff
@@ -46,14 +46,13 @@ describe('Management API:   GET  ->  /correlations/process_model/:process_model_
...
 
-  it('should return a correlation by its id through the Management API', async () => {
+  it('should return a correlation by its id for an user through the Management API', async () => {
 
     const correlations = await testFixtureProvider
       .managementApiClientService
       .getCorrelationsByProcessModelId(testFixtureProvider.identities.defaultUser, processModelId);
 
     should(correlations).be.an.Array();
-    should(correlations.length).be.greaterThan(0);
```
Comment (ElRaptorus) [inline test/2_management_api/correlations/get_by_process_model_id.js]: Why was this removed?

**Comment by ElRaptorus** [comment]
@alexanderkasten The tests you refer to only check if the user has provided an access token with the request, they do not perform any sort of identity-specific request.

File: test/2_management_api/correlations/get_by_process_model_id.js (hunk 3 above)
Comment (alexanderkasten) [reply test/2_management_api/correlations/get_by_process_model_id.js]: because if there are no correlations for a user the `getCorrelationsByProcessModelId` method returns an emtpy array

File: test/2_management_api/correlations/get_all_correlations.js (hunk 2 above)
Comment (alexanderkasten) [reply test/2_management_api/correlations/get_all_correlations.js]: because if there are no correlations for a user the `getCorrelationsByProcessModelId` method returns an emtpy array

File: test/2_management_api/correlations/get_active_correlations.js (hunk 1 above)
Comment (alexanderkasten) [reply test/2_management_api/correlations/get_active_correlations.js]: because if there are no correlations for a user the `getCorrelationsByProcessModelId` method returns an emtpy array

Comment (ElRaptorus) [reply test/2_management_api/correlations/get_active_correlations.js]: But the test specifically starts and finishes a ProcessInstance for this identity, so there must be at least one entry present here.

File: test/2_management_api/correlations/get_all_correlations.js (hunk 2 above)
Comment (ElRaptorus) [reply test/2_management_api/correlations/get_all_correlations.js]: see above

File: test/2_management_api/correlations/get_by_process_model_id.js (hunk 3 above)
Comment (ElRaptorus) [reply test/2_management_api/correlations/get_by_process_model_id.js]: see above

File: test/2_management_api/correlations/get_active_correlations.js (hunk 1 above)
Comment (ElRaptorus) [reply test/2_management_api/correlations/get_active_correlations.js]: Just restore these lines, please 😉

File: test/2_management_api/correlations/get_active_correlations.js [hunk 4]
```diff
@@ -82,11 +92,29 @@ describe('Management API:   GET  ->  /correlations/active', () => {
...
+    correlationListDefaultUser.forEach((correlation) => {
+      should(correlation.identity.userId).equal(defaultIdentity.userId)
+    });
+
+    const correlationListSecondUser = await testFixtureProvider
+      .managementApiClientService
+      .getActiveCorrelations(secondDefaultIdentity);
+
+    correlationListSecondUser.forEach((correlation) => {
+      should(correlation.identity.userId).equal(secondDefaultIdentity.userId)
```
Comment (ElRaptorus) [inline test/2_management_api/correlations/get_active_correlations.js]: Nicely done 👍

File: test/2_management_api/correlations/get_by_process_instance_id.js [hunk 5]
```diff
@@ -111,4 +115,18 @@ describe('Management API:   GET  ->  /correlations/process_instance/:process_ins
     }
   });
 
+  it('should only return correlations by ProcessInstnceId of a specific user', async () => {
```
Comment (ElRaptorus) [inline test/2_management_api/correlations/get_by_process_instance_id.js]:
```suggestion
  it('should only return correlations by ProcessInstanceId for a specific user', async () => {
```

**Comment by ElRaptorus** [comment]
@alexanderkasten Also please consider writing a more comprehensive changelog that we can use with the release. 
`Add Identity to CorrelationService` doesn't really say much. 
Does the service now have an identity attached to it? 

Try describing the UseCases you attempted to fulfill with this, like, for example:
`Refactor Correlation Queries to filter their results by a given identity to ensure that users can only see their own correlations`.

This way we can use the changelogs in the PRS 1:1 for the changelog of the next release.

/cc @robpal94 @SteffenKn @Vyperus

File: test/2_management_api/correlations/get_by_process_model_id.js [hunk 6]
```diff
@@ -38,15 +39,15 @@ describe('Management API:   GET  ->  /correlations/process_model/:process_model_
...
+      .startProcessInstance(identity, processModelId, startEventId, payload, returnOn);
 
     should(result).have.property('correlationId');
     should(result.correlationId).be.equal(payload.correlationId);
 
     return result.correlationId;
   }
 
-  it('should return a correlation by its id through the Management API', async () => {
+  it('should return a correlation by its id for an user through the Management API', async () => {
```
Comment (robpal94) [inline test/2_management_api/correlations/get_by_process_model_id.js]:
```suggestion
  it('should return a correlation by its id for a user through the Management API', async () => {
```

Classic pitfall: https://english.stackexchange.com/questions/105116/is-it-a-user-or-an-user
